crates/factors/src/lib.rs

```diff
@@ -0,0 +1,343 @@
+    ///
+    /// Fails if the current [`SpinFactors`] does not include the given
+    /// [`Factor`] or if the given [`Factor`]'s preparer has not been
+    /// initialized yet (because it is sequenced after this factor).
+    pub fn get_mut<T: Factor>(&mut self) -> Result<&mut T::InstancePreparer> {
+        Factors::instance_preparer_mut::<T>(self.inner)
+            .and_then(|maybe_preparer| maybe_preparer.context("preparer not yet initialized"))
+            .with_context(|| {
+                format!(
+                    "could not get instance preparer for {}",
+                    std::any::type_name::<T>()
+                )
+            })
+    }
+}
+
+pub struct ConfiguredApp<Factors: SpinFactors> {
+    app: App,
+    app_configs: Factors::AppConfigs,
+}
+
+impl<Factors: SpinFactors> ConfiguredApp<Factors> {
+    #[doc(hidden)]
+    pub fn new(app: App, app_configs: Factors::AppConfigs) -> Self {
+        Self { app, app_configs }
+    }
+
+    pub fn app(&self) -> &App {
+        &self.app
+    }
+
+    pub fn app_config<T: Factor>(&self) -> Result<&T::AppConfig> {
+        Factors::app_config::<T>(&self.app_configs).context("no such factor")
+    }
+}
+
+/// Implemented by `#[derive(SpinFactors)]`
+pub trait SpinFactors: Sized {
+    type AppConfigs;
+    type InstancePreparers;
+    type InstanceState: Send + 'static;
+
+    #[doc(hidden)]
+    unsafe fn instance_preparer_offset<T: Factor>() -> Option<usize>;
+
+    #[doc(hidden)]
+    unsafe fn instance_state_offset<T: Factor>() -> Option<usize>;
+
+    fn app_config<T: Factor>(app_configs: &Self::AppConfigs) -> Option<&T::AppConfig>;
+
+    fn instance_state_getter<T: Factor>() -> Option<Getter<Self::InstanceState, T::InstanceState>> {
+        let offset = unsafe { Self::instance_state_offset::<T>()? };
+        Some(Getter {
+            offset,
+            _phantom: PhantomData,
+        })
+    }
+
+    fn instance_state_getter2<T1: Factor, T2: Factor>(
+    ) -> Option<Getter2<Self::InstanceState, T1::InstanceState, T2::InstanceState>> {
+        let offset1 = unsafe { Self::instance_state_offset::<T1>()? };
+        let offset2 = unsafe { Self::instance_state_offset::<T2>()? };
+        assert_ne!(
+            offset1, offset2,
+            "instance_state_getter2 with same factor twice would alias"
+        );
+        Some(Getter2 {
+            offset1,
+            offset2,
+            _phantom: PhantomData,
+        })
+    }
+
+    fn instance_preparer_mut<T: Factor>(
+        preparers: &mut Self::InstancePreparers,
+    ) -> Result<Option<&mut T::InstancePreparer>> {
+        unsafe {
+            let offset = Self::instance_preparer_offset::<T>().context("no such factor")?;
+            let ptr = preparers as *mut Self::InstancePreparers;
+            let opt = &mut *ptr.add(offset).cast::<Option<T::InstancePreparer>>();
+            Ok(opt.as_mut())
+        }
+    }
+}
+
+pub struct Getter<T, U> {
```

Comment:
I think we might want to talk through some of implementation here. There's quite a bit of unsafe code here, and I'm not sure why.

Reply:
Some of this is now gone. Still 2 `transmute`s in the derive code. I briefly tried replacing them with `Any::downcast_ref` but ran into mysterious problems.